mlir/lib/Catalyst/Transforms/DetensorizeSCFPass.cpp

```diff
@@ -0,0 +1,354 @@
+                Value value = rewriter.create<tensor::FromElementsOp>(
+                    newForOp->getLoc(), RankedTensorType::get({}, for_result.getType()),
+                    for_result);
+                rewriter.replaceUsesWithIf(forOp->getResult(index), value, [&](OpOperand &op) {
+                    return !isa<tensor::FromElementsOp>(op.getOwner());
+                });
+            }
+        }
+
+        rewriter.replaceOp(forOp, newForOp);
+        return success();
+    }
+};
+
+struct DetensorizeIfOp : public OpRewritePattern<scf::IfOp> {
+    using OpRewritePattern<scf::IfOp>::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(scf::IfOp ifOp, PatternRewriter &rewriter) const override
+    {
+        // Early exit if there are no results that could be replaced.
+        if (ifOp.getNumResults() == 0) {
+            return failure();
+        }
+        if (!hasScalarTensorResult(ifOp)) {
+            return failure();
+        }
+
+        // 1. Extract tensor elements before yield op
+        SmallVector<scf::YieldOp> yieldOps = {ifOp.thenYield(), ifOp.elseYield()};
+        for (scf::YieldOp yield_op : yieldOps) {
+            // Loop over yield operands
+            for (const auto &it : llvm::enumerate(yield_op.getOperands())) {
+                Value operand = it.value();
+                // Detensorize operand: extract tensor element before yielding
+                if (isScalarTensor(operand)) {
+                    OpBuilder::InsertionGuard g(rewriter);
+                    rewriter.setInsertionPoint(yield_op);
+                    Value value = rewriter.create<tensor::ExtractOp>(yield_op->getLoc(), operand,
+                                                                     ValueRange{});
+                    yield_op.setOperand(it.index(), value);
+                }
+            }
+        };
+
+        // Collect the types for the new IfOp
+        SmallVector<Type> newResultTypes;
+        for (auto result : ifOp->getResults()) {
+            if (isScalarTensor(result)) {
+                newResultTypes.push_back(cast<TensorType>(result.getType()).getElementType());
+                continue;
+            }
+            newResultTypes.push_back(result.getType());
+        }
+
+        // 2. Create the new IfOp
+        OpBuilder::InsertionGuard ifOpInsertionGuard(rewriter);
+        rewriter.setInsertionPoint(ifOp);
+        auto newIfOp =
+            rewriter.create<scf::IfOp>(ifOp.getLoc(), newResultTypes, ifOp.getCondition());
+        rewriter.cloneRegionBefore(ifOp.getThenRegion(), newIfOp.getThenRegion(),
+                                   newIfOp.getThenRegion().end());
+        rewriter.cloneRegionBefore(ifOp.getElseRegion(), newIfOp.getElseRegion(),
+                                   newIfOp.getElseRegion().end());
+
+        // 3. Retensorize results after if op
+        {
+            OpBuilder::InsertionGuard g(rewriter);
+            rewriter.setInsertionPointAfter(ifOp);
```

Comment:
Should this be after the new if op?
```suggestion
            rewriter.setInsertionPointAfter(newIfOp);
```

Reply:
Or even simpler since we are just below the creation of the new if op, the IP should already be there.  Instead we can remove the additional guard and scoped block.